test/functional/p2p_leak_tx.py

```diff
@@ -0,0 +1,93 @@
+        # are more likely to spy on us than outbound peers. Thus, on average,
+        # we announce transactions first to outbound peers, then to (all)
+        # inbound peers. Inbound peers must not be able to successfully request a
+        # transaction if they haven't yet received the announcement for it.
+        #
+        # With only one outbound peer, we expect that a tx is first announced
+        # to (all) inbound peers (and thus present a potential leak) in 28.5% of
+        # the cases.
+        #
+        # Probability( time_ann_inbound < time_ann_outbound )                 =
+        # ∫ f_in(x)                           * F_out(x)                   dx =
+        # ∫ (lambda_in * exp(-lambda_in * x)) * (1 - exp(-lambda_out * x)) dx =
+        # 0.285714
+        #
+        # Where,
+        # * f_in is the pdf of the exponential distribution for inbound peers,
+        #   with lambda_in = 1 / INVENTORY_BROADCAST_INTERVAL = 1/5
+        # * F_out is the cdf of the expon. distribution for outbound peers,
+        #   with lambda_out = 1 / (INVENTORY_BROADCAST_INTERVAL >> 1) = 1/2
+        #
+        # Due to measurement delays, the actual monte-carlo leak is a bit
+        # higher. Assume a total delay of 0.6 s (Includes network delays and
+        # rpc delay to poll the raw mempool)
+        #
+        # Probability( time_ann_inbound < time_ann_outbound + 0.6 )           =
+        # ∫ f_in(x)                           * F_out(x + 0.6)             dx =
+        # ∫ (lambda_in * exp(-lambda_in * x)) * (1 - exp(-lambda_out * (x+.6))) dx =
+        # 0.366485
+        # EXPECTED_MEASURED_LEAK = .366485
+        # Because this test is empirical and our testing framework isn't set up
+        # to handle tests that fail with some expected likelihood, we bump this
+        # value up to decrease the false positive rate.
+        EXPECTED_MEASURED_LEAK = .50
+
+        REPEATS = 100
+        measured_leak = 0
+        self.log.info('Start simulation for {} repeats'.format(REPEATS))
+        for i in range(REPEATS):
+            self.log.debug('Run {}/{}'.format(i, REPEATS))
+            txid = gen_node.sendtoaddress(gen_node.getnewaddress(), 0.033)
+            want_tx = msg_getdata()
+            want_tx.inv.append(CInv(t=1, h=int(txid, 16)))
+
+            wait_until(lambda: txid in outbound_peer.getrawmempool(), lock=mininode_lock)
+            inbound_peer.send_message(want_tx)
+            inbound_peer.sync_with_ping()
+
+            if inbound_peer.last_message.get('notfound'):
+                assert_equal(inbound_peer.last_message['notfound'].vec[0].hash, int(txid, 16))
+                inbound_peer.last_message.pop('notfound')
+            else:
+                measured_leak += 1
+
+        measured_leak /= REPEATS
+        self.log.info('Measured leak of {}'.format(measured_leak))
+
+        assert_greater_than(EXPECTED_MEASURED_LEAK, measured_leak)
```

Comment:
I don't understand what this test is supposed to be testing. Is it:

(i) that we don't leak txs to inbound peers that we haven't yet announced to (as stated in the docstring); or
(ii) that txs are probabilistically announced to outbound peers before inbound peers.

If (i), can't we just test that for every tx, either we receive a `NOTFOUND` message for the tx or we've already received an `INV` message for the tx?

The `assert_greater_than()` call here seems to be simply testing (ii). Or maybe I'm misunderstanding?

Reply:
Concept ACK the change to the test!

Can we add multiple P2P connections to the node to get more test repeats in parallel, or does the change to tx propagation in https://github.com/bitcoin/bitcoin/pull/13298 mean that we'll announce to all those peers at the same time and not get any additional benefit?

Reply:
Currently we can only easily add inbound mininode peers (which are in the same bucket for tx propagation), so right now it couldn't run in parallel that way.

I am happy to adjust the test as soon as we can add outbound connections to mininode peers.

Reply:
Thanks!